### ISSUE TITLE
Allow to specify the salt length for PKCS#1 RSASSA PSS signature generation

### ISSUE THREAD

**Issue by meuter** [Mbed-TLS/mbedtls#3152]
## Description
- Type: Feature Request
- Priority: Minor

----------------------------------------------------------------
## Feature Request

I'm trying to run the CAVP test vectors for PKCS#1 RSASSA PSS signature generation (can be found [here](https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Digital-Signatures#rsa2vs)). And I'm running into a small problem with this function:

```c
int mbedtls_rsa_rsassa_pss_sign( mbedtls_rsa_context *ctx,
                         int (*f_rng)(void *, unsigned char *, size_t),
                         void *p_rng,
                         int mode,
                         mbedtls_md_type_t md_alg,
                         unsigned int hashlen,
                         const unsigned char *hash,
                         unsigned char *sig );
```
The problem is that the test vector provides the salt value and expects the generated signature to be equal to a precomputed one. Here's an example:

```
n = bcb47b2e0dafcba81ff2a2b5cb115ca7e757184c9d72bcdcda707a146b3b4e29989ddc660bd694865b932b71ca24a335cf4d339c719183e6222e4c9ea6875acd528a49ba21863fe08147c3a47e41990b51a03f77d22137f8d74c43a5a45f4e9e18a2d15db051dc89385db9cf8374b63a8cc88113710e6d8179075b7dc79ee76b
e = 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010001
d = 383a6f19e1ea27fd08c7fbc3bfa684bd6329888c0bbe4c98625e7181f411cfd0853144a3039404dda41bce2e31d588ec57c0e148146f0fa65b39008ba5835f829ba35ae2f155d61b8a12581b99c927fd2f22252c5e73cba4a610db3973e019ee0f95130d4319ed413432f2e5e20d5215cdd27c2164206b3f80edee51938a25c1
SHAAlg = SHA1
SaltVal = 6f2841166a64471d4f0b8ed0dbb7db32161da13b
Msg = 1248f62a4389f42f7b4bb131053d6c88a994db2075b912ccbe3ea7dc611714f14e075c104858f2f6e6cfd6abdedf015a821d03608bf4eba3169a6725ec422cd9069498b5515a9608ae7cc30e3d2ecfc1db6825f3e996ce9a5092926bc1cf61aa42d7f240e6f7aa0edb38bf81aa929d66bb5d890018088458720d72d569247b0c
S = 682cf53c1145d22a50caa9eb1a9ba70670c5915e0fdfde6457a765de2a8fe12de9794172a78d14e668d498acedad616504bb1764d094607070080592c3a69c343d982bd77865873d35e24822caf43443cc10249af6a1e26ef344f28b9ef6f14e09ad839748e5148bcceb0fd2aa63709cb48975cbf9c7b49abc66a1dc6cb5b31a
```
To inject the salt value, I can probably use `f_rng` (although I did not try yet - but I don't see any reason it would not work), but for the salt length, however, it is computed inside the function itself:

```c
    hlen = mbedtls_md_get_size( md_info );

    /* Calculate the largest possible salt length. Normally this is the hash
     * length, which is the maximum length the salt can have. If there is not
     * enough room, use the maximum salt length that fits. The constraint is
     * that the hash length plus the salt length plus 2 bytes must be at most
     * the key length. This complies with FIPS 186-4 §5.5 (e) and RFC 8017
     * (PKCS#1 v2.2) §9.1.1 step 3. */
    min_slen = hlen - 2;
    if( olen < hlen + min_slen + 2 )
        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
    else if( olen >= hlen + hlen + 2 )
        slen = hlen;
    else
        slen = olen - hlen - 2;
```

Unfortunately, some of the CAVP tests use a salt that is smaller or bigger than hlen. And for those test the generated signature does not match (which is normal since the salt used has a different length). 

For comparison, I had a quick look at BoringSSL does and the corresponding function does allow to inject the salt length:

```c
int RSA_sign_pss_mgf1(RSA *rsa, size_t *out_len, uint8_t *out, size_t max_out,
                      const uint8_t *in, size_t in_len, const EVP_MD *md,
                      const EVP_MD *mgf1_md, int salt_len)
```

**Justification - why does the library need this feature?**  

We need to provide evidence that our crypto toolbox passes CAVP tests vectors for regulatory purposes (some security certification).

**Suggested enhancement**  

Would you be open to allow the salt length to be provided? One way to do it would be to have an `_ext` function that takes the `salt_len` as input, like it's done for the verification:

```c
int mbedtls_rsa_rsassa_pss_verify_ext( mbedtls_rsa_context *ctx,
                               int (*f_rng)(void *, unsigned char *, size_t),
                               void *p_rng,
                               int mode,
                               mbedtls_md_type_t md_alg,
                               unsigned int hashlen,
                               const unsigned char *hash,
                               mbedtls_md_type_t mgf1_hash_id,
                               int expected_salt_len,   ///// <<<< here the salt length can be provided
                               const unsigned char *sig )
```
Another options would be to store the salt length in the RSA context like is done for the padding (V15 vs V21) and have a default value that would fallback to using the hash length. This may be less invasive.

With a little help, I'd willing to contribute a patch, but before jumping in, I'd would appreciate your feedback, maybe there's another solution that eludes me at this point. :-)

Thanks!

**Comment by gilles-peskine-arm** [comment]
Hello,

I didn't realize that CAVP for RSA-PSS required support for smaller salts. That's a pretty strong argument to support PSS. I was aware that X.509 allows variable salt size but X.509 with PSS isn't very common so that didn't seem so important.

Both adding a `pss_verify_ext` similar to `pss_sign_ext`, and making the salt size part of the context, make sense. It depends whether the salt length should be considered part of a key's policy (always use a given salt size with a given key) or not (a key may be used with any salt size). Since the hash identifier is part of the context, I guess the salt size should also be.

Annoyingly, the type `mbedtls_rsa_context` is semi-public: the documentation states that “Direct manipulation of the members of this structure is deprecated”, but not outright forbidden. So the layout of the structure is technically part of the library's public API. On the other hand, looking inside the context is not allowed if the application might support RSA accelerators, since alternative implementation of RSA can provide their own implementation of the type `mbedtls_rsa_context`. So I think it's acceptable to change the context type.

It would also be possible to extend the meaning of the `int padding` field in the context, which currently only allows `MBEDTLS_RSA_PKCS_V15` and `MBEDTLS_RSA_PKCS_V21`. We could add `MBEDTLS_RSA_PKCS_V21_WITH_SALT(salt)`. But this may be a bit confusing.

Your contribution would be welcome. But before that I'd like to hear the opinion of other members of the Mbed TLS team.

The PSA API also only supports signing (and even verification) only with salt size = hash size. This needs to be extended too; tracked internally at Arm at https://github.com/ARMmbed/psa-crypto-api/issues/58.

**Comment by gilles-peskine-arm** [comment]
Injecting the salt via `f_rng` should work. But it's fragile: the same RNG is also used for blinding. Luckily in this particular case the salt generation happens before blinding (and you can't really do it any other way).

**Comment by mpg** [comment]
I also think that both providing a `sign_ext` function and extending the context are viable options. As Gilles says, since the hash is part of the context, it would probably make more sense to the salt length to be there as well.

I don't see a compatibility problem with adding a new member to that context in the development branch.

**Comment by meuter** [comment]
Thanks for your feedback. So if that's OK for everyone, I propose to work on a `sign_ext` function then.

**Comment by meuter** [comment]
I just submitted a first draft of the `_sign_ext()` solution. Let me know what you think.

**Comment by daverodgman** [comment]
Fixed by #3183
